[Persistence] Followup on PrePersistence deprecation & Block lifecycle

Olshansk opened the pull request:
# Objective

Follow up on both critical and non-critical fixes to https://github.com/pokt-network/pocket/pull/140.

# Origin Document

https://github.com/pokt-network/pocket/pull/140 deprecated PrePersistence and implemented multiple contexts. This PR is a follow with some stylistic changes, improvements in the readability of the tests, the addition of unit tests, bug fixes in some SQL business logic, and implementation of the read context.

Followup work is being tracked in https://github.com/pokt-network/pocket/issues/149.

## Learnings to streamline workflow in the future
- Do not conflate non-critical stylistic changes as it leads to more conflicts down the road
- Implemented an automated "LocalNet" for easier testing
- Try to plan work ahead of time so we can work in smaller PRs / iterations

## Changelog / changes

**Main persistence module changes:**

- Split `ConnectAndInitializeDatabase` into `connectToDatabase` and `initializeDatabase`
  - This enables creating multiple contexts in parallel without re-initializing the DB connection
- Fix the SQL query used in `SelectActors`, `SelectAccounts` & `SelectPools`
  - Add a generalized unit test for all actors
- Remove `NewPersistenceModule` and an injected `Config` + `Create`
  - This improves isolation a a “injection-like” paradigm for unit testing
- Change `SetupPostgresDocker` to `SetupPostgresDockerPersistenceMod`
  - This enables more “functional” like testing by returning a persistence module and avoiding global testing variables
  - Only return once a connection to the DB has been initialized reducing the likelihood of test race conditions
- Implemented `NewReadContext` with a proper read-only context

**Secondary persistence module changes**

- Improve return values in `Commit` and `Release` (return error, add logging, etc…)
- Add `pgx.Conn` pointer to `PostgresDB`
- `s/db/conn/g` and `s/conn/tx/g` in some (not all) places where appropriate
- Make some exported variables / functions unexported for readability & access purposes
- Add a few helpers for persistence related unit testing
- Added unit tests and TODOs for handling multiple read/write contexts

## General milestone checklist
- [x] Update all the relevant CHANGELOGs
- [ ] Update all the relevant READMEs
- [ ] Update the source code tree explanation
- [ ] Add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [x] Create followup milestones + issues
- [x] Document small TODO along the way

## Follow-up Work

All follow-up work is being tracked in https://github.com/pokt-network/pocket/issues/149

---

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] If applicable, I have made corresponding changes to related or global README
- [ ] If applicable, I have added added new diagrams using [memaid.js](https://mermaid-js.github.io)
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works

Commit:
[Persistence] First iteration of a PostgreSQL based Persistence Schema (#73)

## Objective

Foundational iteration of PostgreSQL based persistence module implementation.

## Origin Document

https://github.com/pokt-network/pocket/issues/68

## Type of change

New major module implementation.

### Persistence-related core changes:
- List of actors / interfaces with MVP implementation:
    - Applications
    - Fisherman
    - ServiceNode
    - Accounts
    - Pools
- List of actors / interfaces with partial MVP implementation:
    - Validator
    - Gov params
- List of actors / interfaces with minorimplementation:
    - Block
- SQL Schema definition of the actors above
- SQL Query implementation for common actor persistence functionality
- PostgresContext implementation of the actors actors above
- Base infrastructure for fuzz testing

Non-persistence “fly-by” changes
- Updates to the PrePersistence module and utility module with breaking changes
- A few minor improvements/additions to the Makefile
- TODOs & comment cleanups throughout the codebase

## How Has This Been Tested?

### Unit Tests

```
make test_persistence
make test_all
```

### LocalNet
Ran a basic LocalNet following the instructions in the [development README](docs/development/README.md).


Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>
Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MacBook-Pro-2.local>
Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MBP-2.lan>
Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

Olshansk committed:
Merge with main and fix tests

Commit:
only committing due to merge, wip

Commit:
Merge remote-tracking branch 'origin/origin/104-persistence-v1-persistence-foundation-interim-merge' into issue-#111

Commit:
issue-#111

Commit:
merged into main

Commit:
see pull request changes on #111

Commit:
using transaction architecture for persistence module, begin to swith to read / write separation

Commit:
deprecate pre-persistence

Commit:
merged with main

Olshansk committed:
Merge branch 'main' into issue-#128

Olshansk committed:
Remove lingering pre2p files

Olshansk committed:
Minor changelog update

Olshansk committed:
Move major var to the top of file

Olshansk committed:
Minor code cleanup across multiple files

Olshansk committed:
Replaced log.Fatal(fmt.Sprintf with log.Fatalf

Olshansk committed:
Minor updates

Commit:
requested changes 1 for pull/140

Commit:
merge with main

Commit:
requested changes 2 for pull/140

Commit:
Merge remote-tracking branch 'origin/main' into issue-#128

Olshansk committed:
pull and resolve

Olshansk committed:
Replace staking status numbers with iota

Olshansk committed:
Added a TODO

Olshansk committed:
Cleaned up a few tests - but still a WIP

Olshansk committed:
Add a default constant for account amount

Olshansk committed:
Reviewed persistence/genesis.go

Olshansk committed:
Checkpoint commit - removed NewPersistenceModule but some tests still failing

Olshansk committed:
Updated .gitignore

Olshansk committed:
Updated Makefile - remove old unused targets and reorder existing ones

Olshansk committed:
Reverted UndefinedStakingStatus to be explicit

Olshansk committed:
Updated a few comments and TODOs

Olshansk committed:
Checkpoint - got persistence tests passing again

Olshansk committed:
s/SetParam/SetParamQuery and 'make test_all' passes

Olshansk committed:
Added tests for TestGetAllServiceNodes

Olshansk committed:
Created a general purpose getAllActorsTest

Olshansk committed:
All tests pass except for GetAllValidators

Olshansk committed:
All tests passing

Olshansk committed:
All tests passing after GetAllAccounts and GetAllPools is fixed

Olshansk committed:
Review shared/tests/utility_module/transaction_test.go

Olshansk committed:
Minor Makefile changes

Olshansk committed:
Rename SetupPostgresDocker to SetupPostgresDockerPersistenceMod

Olshansk committed:
Misc cleanup

Olshansk committed:
Improved test setup - TestGetAll and ServiceNode tests work

Olshansk committed:
Temp commit

Olshansk committed:
Added a comment

Olshansk committed:
Got 'make test_persistence' to pass

Olshansk committed:
s/ClearAllDebug/DebugClearAll

andrewnguyen22 commented on shared/tests/utility_module/block_test.go:
```diff
@@ -72,11 +79,14 @@ func TestUtilityContext_BeginBlock(t *testing.T) {
-	//require.NoError(t, err)
-	//require.False(t, missed != 1, fmt.Sprintf("wrong missed blocks amount; expected %v got %v", 1, byzantine.MissedBlocks))
-	ctx.Context.Release() // TODO (team) need a golang specific solution for teardown
+	// missed, err := ctx.GetValidatorMissedBlocks(byzantine.Address) TODO not yet implemented
```
See TODO (not yet implemented)

Olshansk replied:
Removed the `DISCUSS_IN_THIS_COMMIT`

andrewnguyen22 commented on persistence/schema/shared_sql.go:
```diff
@@ -119,6 +121,8 @@ func Insert(
 							  height=EXCLUDED.height`,
 		tableName, actorSpecificParam,
 		actor.Address, actor.PublicKey, actor.StakedTokens, actorSpecificParamValue,
+		// DISCUSS_IN_THIS_COMMIT: Why did we change this?
```
I don't have a good answer for this, revert?

Olshansk replied:
I reverted it and it broke the unit tests. Going to investigate

Olshansk replied:
Fixed: https://github.com/pokt-network/pocket/pull/165/commits/7148c17685b28f0734139447c19550712305ea2f

andrewnguyen22 commented on persistence/shared_sql.go:
```diff
@@ -4,13 +4,15 @@ import (
 )
 
-// IMPROVE(team): Move this into a proto enum
+// IMPROVE(team): Move this into a proto enum. We are not using `iota` for the time being
```
(Non-Actionable) StakeStatus is deprecated at the state level in #154 expect more cleanup to come here

Olshansk replied:
ACK

andrewnguyen22 commented on persistence/test/account_test.go:
```diff
@@ -294,13 +303,18 @@ func newTestAccount(t *testing.T) typesGenesis.Account {
 	}
 	return typesGenesis.Pool{
-		Name: DefaultPoolName,
+		Name: fmt.Sprintf("%s_%d", DefaultPoolName, rand.Int()),
```
(Non-Actionable) Pools become proto-enum in #154

Olshansk replied:
ACK

andrewnguyen22 commented on persistence/test/application_test.go:
```diff
@@ -246,6 +236,16 @@ func TestGetSetStakeAmount(t *testing.T) {
+	db := NewTestPostgresContext(t, 0)
+
+	updateApp := func(db *persistence.PostgresContext, app *genesis.App) error {
+		return db.UpdateApp(app.Address, "https://olshansky.info", app.MaxRelays, []string{"OLSH"})
```
NIT: Can make these constants and global var
`https://olshansky.info`
`[]string{"OLSH"}`

Olshansk replied:
Any objections to naming it `OlshanskyURL` and `OlshanskyChain`

andrewnguyen22 replied:
Lol no objection

andrewnguyen22 commented on persistence/test/generic_test.go:
```diff
@@ -34,6 +36,92 @@ func NewTestGenericActor[T any](protocolActorSchema schema.ProtocolActorSchema,
+	require.NoError(t, err)
+	require.Len(t, actors, initialCount+3)
+
+	// DISCUSS_IN_THIS_COMMIT: Since we do not support `DeleteActor`, should we filter here based on status? If so, tests need to be updated.
```
(Non-Actionable) 'Delete' is in limbo for now. I say let's document/leave it

Olshansk replied:
Comment updated.

andrewnguyen22 commented:
First pass - here are my thoughts:
- Most of it is cleanup (not a bad thing)
- There will be a ton of conflicts with #154 (a bad thing)

Per comments in #128 let's talk offline about merge strategy. 

Could get quite messy given all of the outstanding things

Olshansk committed:
Use NewTestPostgresContext everywhere

andrewnguyen22 commented:
> EDIT: Moving this to https://github.com/pokt-network/pocket/pull/165.
>
> I'm updating this to use DISTINCT and avoiding the IN query. Even though this probably yields some performance gains (did not benchmark), I think it's more readable / understandable to see what query we're making.
>
> Also going to add some unit tests here too.

> EDIT: Moving this to https://github.com/pokt-network/pocket/pull/165.
> 
> I added a few unit tests and there seemed to be a bug in the previous approach but I didn't dive deep enough to understand the reasoning. Here is the new business logic:
> 
> func SelectAccounts(height int64, tableName string) string {
```
	return fmt.Sprintf(`
			SELECT DISTINCT ON (address) address, balance, height
			FROM %s
			WHERE height<=%d
			ORDER BY address, height DESC
       `, tableName, height)
}
```

The original approach
```
			SELECT address, balance, height
			FROM %s WHERE height<=%d AND (height,address) IN (SELECT MAX(height),address from %s GROUP BY address)
```

(attempts to) ensure that the latest height is selected from the set.

Can you point me to documentation that shows that `DISTINCT ON` will always select the `first row` from the `ORDER BY` thus guaranteeing that the latest height is always selected.

Olshansk committed:
Added todo_this_commit

Olshansk committed:
Checkpoint commit - test_utility_module passes and test_persistence fails

Olshansk commented:
@andrewnguyen22 While this is still a WIP, can you please take a look at all the `DISCUSS_IN_THIS_COMMIT` comments?

If you pull the branch locally, you can list of all of in terminal using `make todo_this_commit` :)

Olshansk committed:
Fixed how we reference shared variables in tests

Olshansk committed:
Temp commit

Olshansk committed:
Merge with main

Olshansk committed:
Unit tests passing

Olshansk committed:
Split connect and initialize DB into two

Olshansk committed:
Started doing a self review of some minor nits

Olshansk committed:
Finished first round of self-review

Olshansk committed:
Updated the changelog

Olshansk commented:
@andrewnguyen22 This is still incomplete but I made some progress. Most of the remaining action items can be seen with `make todo_this_commit`.

Olshansk committed:
Move olshansky URL and olshansky chains to constants

Olshansk committed:
Remove the used of unused height in sql schemas

andrewnguyen22 commented on persistence/test/application_test.go:
```diff
@@ -209,26 +190,25 @@ func newTestApp() (*typesGenesis.App, error) {
 	}, nil
 }
 
+// DISCUSS_IN_THIS_COMMIT: We are only calling these functions and tests for apps, but need to
```
Okay, is this a discussion?

Olshansk replied:
Sorry, this was mean to be a `TODO_IN_THIS_COMMIT`

andrewnguyen22 commented on persistence/test/setup_test.go:
```diff
@@ -40,38 +39,62 @@ var (
+	}
+	t.Cleanup(func() {
+		require.NoError(t, db.Release())
+		// DISCUSS_IN_THIS_COMMIT: Do we need this given that `Release` reverses the changes made to the database?
```
I think it's cleaner and safer IMO to restart the whole stack. But not too strong on it

Olshansk replied:
I'm going to remove it so that it's easier for us to "run into errors" as we add more changes with unexpected circumstances.

andrewnguyen22 commented on persistence/test/setup_test.go:
```diff
@@ -40,38 +39,62 @@ var (
+	}
+	f.Cleanup(func() {
+		db.Release()
+		// DISCUSS_IN_THIS_COMMIT: Do we need this given that `Release` reverses the changes made to the database?
```
See comment above

andrewnguyen22 commented on utility/types/message.go:
```diff
@@ -18,6 +18,7 @@ import (
 validations possible before even checking the state storage layer.
 */
 
+// DISCUSS_IN_THIS_COMMIT: What's the best location for these shared constants?
```
Not sure

Olshansk committed:
Fix for defaultAccountAmountStr

Olshansk commented on persistence/account.go:
```diff
@@ -10,6 +10,13 @@ import (
 // TODO(https://github.com/pokt-network/pocket/issues/102): Generalize Pool and Account operations.
 
+// DISCUSS_IN_THIS_COMMIT: if we want to make use of `defaultAccountAmountStr`, then we should
+//                         check if the error returned by the DB is `RowNotFound` and return a nil
```
I'm using `pgx.ErrNoRows` for the check which I think is reasonable given that it is associated with the`PostgresContext` struct, so a different platform / struct would check for a different error.

Also added 2 unit test to verify this moving forward. [`3e54025` (#165)](https://github.com/pokt-network/pocket/pull/165/commits/3e54025dce6d0c10699ae35873316587142e145e)

Olshansk commented on persistence/db.go:
```diff
@@ -18,75 +18,74 @@ const (
+	schema.ValidatorActor,
+}
+
+// DISCUSS_IN_THIS_COMMIT: Do we still need this?
```
The rand pacakge is small and we're using it elsewhere so 100% not the problem.

I'm just not sure what we're seeding for here in the persistence module. If anything, we can do it in `node.go`?

tl;dr Tests still passed after removing it, but don't object to adding it back in.

Olshansk commented on persistence/genesis.go:
```diff
@@ -25,67 +30,85 @@ func (pm *persistenceModule) PopulateGenesisState(state *genesis.GenesisState) {
+		}
+	}
+
+	// DISCUSS_IN_THIS_COMMIT: Do protocol actors need corresponding accounts?
```
Let's sync on this offline.

Olshansk replied:
tl;dr When an unstake happens, it is set to 0.

Olshansk replied:
tl;dr When an unstake happens, the amount is set appropriately. Otherwise, it defaults to 0.

Olshansk commented on persistence/module.go:
```diff
@@ -14,56 +15,50 @@ var _ modules.PersistenceModule = &persistenceModule{}
+	}
+
+	// DISCUSS_IN_THIS_COMMIT: Is `Create` the appropriate location for this or should it be `Start`?
+	// DISCUSS_IN_THIS_COMMIT: Thoughts on bringing back `shouldHydrateGenesisDb`? It allowed LocalNet
```
Sounds good. Let's ignore it in this commit.

Olshansk commented on persistence/module.go:
```diff
@@ -14,56 +15,50 @@ var _ modules.PersistenceModule = &persistenceModule{}
+		blockStore:  blockStore,
+	}
+
+	// DISCUSS_IN_THIS_COMMIT: Is `Create` the appropriate location for this or should it be `Start`?
```
Leaving here to reduce PR scope and adding this comment:

```
	// TECHDEBT: reconsider if this is the best place to call `populateGenesisState`. Note that
	// this forces the genesis state to be reloaded on every node startup until state sync is
	// implemented.
```

Olshansk committed:
Reply to several comments

Olshansk committed:
Added a general getTestGetSetStakeAmountTest and added unit tests for each actor

Olshansk committed:
Added more TODOs and unit tests for multiple contexts

Olshansk committed:
Update todo_in_this_commit make target

Olshansk commented:
@andrewnguyen22 Please take a look at this PR and lmk if there any additional changes you'd like for me to make.

andrewnguyen22 commented on persistence/test/module_test.go:
```diff
@@ -1,35 +1,129 @@
+// - keeping a set of all the write contexts and panicking
+// - Adding timeouts to the write contexts
+// - ???
+func TestTwoWriteContextsSameHeight(t *testing.T) {
```
These tests don't make sense to me. As discussed, we should never have two write contexts in parallel

Olshansk replied:
Removed tests and am now erroring on multiple write contexts.

andrewnguyen22 commented on persistence/gov.go:
```diff
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"encoding/hex"
+
```
Why the newline here?

Olshansk replied:
Go imports are automatically formatted by my vscode Golang extension based on the "latest best practices".

For me it's an auto-edit when I do a cmd+save.

I looked online (https://stackoverflow.com/a/38714480/768439) and seems like the best practice that the linter implements is grouping by package (e.g. github vs std) separated by a new line.

<img width="755" alt="Screen Shot 2022-08-17 at 11 38 54 AM" src="https://user-images.githubusercontent.com/1892194/185182611-8d3cc95a-1c5b-4a07-96fe-bc94945be83c.png">

andrewnguyen22 reviewed (changes requested):
All looks good. 99% there. Just two comments

Olshansk committed:
Error on multiple write contexts

Olshansk committed:
Remove deprecated tests

Olshansk committed:
Merge with main

Olshansk committed:
Followup fixes after merge with main - unit tests still failing

Olshansk committed:
Use a read context in consensus tests when loading

Olshansk committed:
Fixed the block lifecycle after testing with LocalNet

Olshansk committed:
Added hack to clear previous state in genesis state hydration

Olshansk commented:
@andrewnguyen22 Please take a close look at the commits between HEAD and `f9ddc0954de45f09559d7812de6e8b8857e9dc2d`. All the unit tests are passing but I also had to make several block lifecycle modifications for LocalNet to succeed.

Olshansk committed:
Populate genesis state in DEBUG_CLEAR_STATE

andrewnguyen22 commented on consensus/module.go:
```diff
@@ -172,8 +172,9 @@ func (m *consensusModule) loadPersistedState() error {
 	// TODO: Populate the rest of the state from the persistence module: validator set, quorum cert, last block hash, etc...
-	m.Height = uint64(latestHeight) + 1 // +1 because the height of the consensus module is where it is actively participating in consensus
-	m.appHash = string(appHash)
+	// m.Height = uint64(latestHeight) + 1 // +1 because the height of the consensus module is where it is actively participating in consensus
```
Why is this left as comments... 

What's going on with the appHash? Just logging?

Olshansk replied:
Adding back w/ previous state hydration.

andrewnguyen22 commented on persistence/genesis.go:
```diff
@@ -2,21 +2,33 @@ package persistence
+func (m *persistenceModule) populateGenesisState(state *genesis.GenesisState) {
+	log.Println("Populating genesis state...")
+
+	// HACK: This is needed to avoid block a previous from interfering with the genesis state hydration
```
This is very unclear to me. Populate Genesis should only be called once

andrewnguyen22 replied:
This is non-actionable, but this comment speaks to how confusing the entire lifecycle issue is to me. I am not clear on the bug nor the solution at this time

andrewnguyen22 commented on persistence/gov.go:
```diff
@@ -27,6 +27,7 @@ func (p PostgresContext) InitParams() error {
 	if err != nil {
 		return err
 	}
+
```
Why newline?

andrewnguyen22 commented on persistence/module.go:
```diff
@@ -95,18 +123,31 @@ func (m *persistenceModule) NewRWContext(height int64) (modules.PersistenceRWCon
-	return m.NewRWContext(height)
-	// TODO (Team) this can be completely separate from rw context.
-	// It should access the db directly rather than using transactions
+func (m *persistenceModule) ResetContext() error {
```
This appears to only be a test function and not part of the lifecycle? 

Please add TODO to make it a test func only

andrewnguyen22 replied:
There's nothing I see here stoping this from being at the context level

andrewnguyen22 commented on shared/modules/persistence_module.go:
```diff
@@ -9,6 +9,7 @@ type PersistenceModule interface {
 	Module
 	NewRWContext(height int64) (PersistenceRWContext, error)
 	NewReadContext(height int64) (PersistenceReadContext, error)
+	ResetContext() error
```
TODO remove from interface once moved to test function and/or re-scope it at the context level

Olshansk replied:
Added.

andrewnguyen22 commented:
`Utility context not nil when preparing a new block? Releasing for now but should not happen`

Olshansk commented:
@andrewnguyen22  comment regarding the genesis / lifecycle piece. 

When we added the blockstore in https://github.com/pokt-network/pocket/pull/121 , we also added a `shouldHydrateGenesis` function so that:
1. We don't run `populateGenesis` if there is already a previous state
2. Pick up the previous state if we already had one (hence the logic around loading the previous hash and height which I'm just logging)

https://github.com/pokt-network/pocket/pull/140/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.mod&file-filters%5B%5D=.sum&file-filters%5B%5D=No+extension&show-viewed-files=true&show-deleted-files=false

<img width="654" alt="Screen Shot 2022-08-18 at 6 45 18 AM" src="https://user-images.githubusercontent.com/1892194/185377180-e05adc91-f5d6-44f3-8167-61ee78a188f7.png">

My understanding is that per the other comments / discussions (screenshot below), we didn't want to bring it back in favor of the future configs changes happening, but I had to add a couple stop-gap hacks in the meantime so that we load up the genesis correctly every time.

I'd personally prefer to add it back in this PR so there are less hacks / confusion, but also unsure per you your other comments. Lmk how you think I should proceed.

<img width="809" alt="Screen Shot 2022-08-18 at 6 46 21 AM" src="https://user-images.githubusercontent.com/1892194/185377178-64f7249a-3e8a-4318-85c1-06f2efab00d4.png">

Olshansk committed:
Remove the hack for loading a previous state

Olshansk committed:
Update comment

Olshansk committed:
Added a comment explaining refreshUtilityContext
